tests/test_api_render.py

```diff
@@ -299,3 +301,35 @@ def test_pin_expression_works_with_python_prereleases(testing_config):
     assert len(ms) == 2
     m = next(m_[0] for m_ in ms if m_[0].meta["package"]["name"] == "bar")
     assert "python >=3.10.0rc1,<3.11.0a0" in m.meta["requirements"]["run"]
+
+
+@pytest.mark.benchmark
+def test_pin_subpackage_benchmark(testing_config):
+    # Performance regression test for https://github.com/conda/conda-build/pull/5224
+    recipe = os.path.join(metadata_dir, "_pin_subpackage_benchmark")
+
+    # Create variant config of size comparable (for subdir linux-64) to
+    #   https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/3c7d60f56a8cb7d1b8f5a8da0b02ae1f1f0982d7/recipe/conda_build_config.yaml
+    # Addendum: Changed number of single-value keys from 327 to 33 to reduce benchmark duration.
+    def create_variant():
+        # ("pkg_1, ("1.1", "1.2", ...)), ("pkg_2", ("2.1", "2.2", ...)), ...
+        packages = ((f"pkg_{i}", (f"{i}.{j}" for j in count(1))) for i in count(1))
+        variant = {}
+        variant["zip_keys"] = []
+        for version_count, package_count in [(1, 4), (4, 3), (4, 3)]:
+            zipped = []
+            for package, versions in islice(packages, package_count):
+                zipped.append(package)
+                variant[package] = list(islice(versions, version_count))
+            variant["zip_keys"].append(zipped)
+        # for version_count, package_count in [(3, 1), (2, 4), (1, 327)]:
+        for version_count, package_count in [(3, 1), (2, 4), (1, 33)]:
+            for package, versions in islice(packages, package_count):
+                variant[package] = list(islice(versions, version_count))
+        validate_spec("<generated>", variant)
+        return variant
+
+    ms = api.render(
+        recipe, config=testing_config, channels=[], variant=create_variant()
```

Comment:
Wow `variant` versus `variants`, what a deep rabbit hole...

Reply:
Oh, yeah, maybe we do `variants=[create_variant()]` instead? I think `variant` might be some legacy thing (haven't checked though).

Reply:
`variants` is also a dict, using `variants=[create_variant()]` will error

Reply:
I think `Config.variants` is supposed to be a list, equal to `[config.variant]` in case there is only one.
(But I'd have to check; and also check whether we'd rather want `variants` or not to avoid re-using old stuff.)

Reply:
Oh, now I get what you meant.
`render` also takes a `variants` argument, so that one wouldn't end up as `config.variants` then.
Plus, I also think that just using `api.render(..., variants=create_variants())` would be the correct thing to do since I think `Config.variant`/`Config.variants` would be the flattened/"exploded" variants (i.e., every key is supposed to have a single value; using `render(..., variant=...)` probably only worked because the test don't actually use the entries).

Reply:
Changed it to `variants`; let's see how much it affects the run time (since that way we should get multiple entries in `Config.variants`).

Reply:
> Changed it to `variants`; let's see how much it affects the run time (since that way we should get multiple entries in `Config.variants`).

Didn't change a thing, which is good to know!